### DISCUSSION TITLE
toActionObject is slow due to Object.defineProperty -- Can it be replaced?

### DISCUSSION THREAD

**Discussion by theRobinator** [statelyai/xstate#2488] [Q&A]
I work on a relatively-good-sized website and we're using XState on a large portion of our pages. We do heavy caching, but we still end up rendering pages with XState on them about 10 times per second all day. Today I was looking into how to speed up page rendering, and I ended up running the Node profiler on an instance. It showed me this:

![image](https://user-images.githubusercontent.com/1115313/127934610-35549954-e4b7-4a55-99d3-6add65c60771.png)

As you can see, `toActionObject` is taking more than a third of the total time for rendering a page. When I looked into it further, I found this:
![image](https://user-images.githubusercontent.com/1115313/127934661-1d3d12d7-9c4b-4971-9965-f9b7b5dbc88f.png)

It's just `Object.defineProperty` that takes up all that time! A bit of Googling showed me [that this can be made faster by simply setting the property](https://humanwhocodes.com/blog/2015/11/performance-implication-object-defineproperty/) instead of using `Object.defineProperty` -- but then `toString` would be enumerable.

So my question is, **is it safe to make this change?** If so, I'm happy to submit a PR.

**Comment by davidkpiano** [comment]
Added https://github.com/statelyai/xstate/pull/2491 to fix this 👍

**Reply by theRobinator** [reply]
Thanks so much @davidkpiano!

**Comment by ChrisShank** [comment]
Super curious if your solution have seen any other bottlenecks with `xstate`?
